localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp

```diff
@@ -819,68 +820,67 @@ std::array<double, 36> NDTScanMatcher::estimate_covariance(
-    sub_initial_pose_matrix(1, 3) += static_cast<float>(rotated_pose_offset_2d.y());
-
-    auto sub_output_cloud = std::make_shared<pcl::PointCloud<PointSource>>();
-    ndt_ptr_->align(*sub_output_cloud, sub_initial_pose_matrix);
-    const Eigen::Matrix4f sub_ndt_result = ndt_ptr_->getResult().pose;
-
-    const Eigen::Vector2d sub_ndt_pose_2d = sub_ndt_result.topRightCorner<2, 1>().cast<double>();
-    mean += sub_ndt_pose_2d;
-    ndt_pose_2d_vec.emplace_back(sub_ndt_pose_2d);
-
-    multi_ndt_result_msg.poses.push_back(matrix4f_to_pose(sub_ndt_result));
-    multi_initial_pose_msg.poses.push_back(matrix4f_to_pose(sub_initial_pose_matrix));
-  }
+  Eigen::Matrix2d estimated_cov_2d = Eigen::Matrix2d::Zero();
+  estimated_cov_2d.diagonal() << ndt_covariance[0 + 6 * 0], ndt_covariance[1 + 6 * 1];
 
-  // calculate the covariance matrix
-  mean /= n;
-  Eigen::Matrix2d pca_covariance = Eigen::Matrix2d::Zero();
-  for (const auto & temp_ndt_pose_2d : ndt_pose_2d_vec) {
-    const Eigen::Vector2d diff_2d = temp_ndt_pose_2d - mean;
-    pca_covariance += diff_2d * diff_2d.transpose();
+  if (
+    param_.covariance.covariance_estimation.covariance_estimation_type ==
+    CovarianceEstimationType::LAPLACE_APPROXIMATION) {
+    Eigen::Matrix2d cov_by_la =
+      pclomp::estimate_xy_covariance_by_Laplace_approximation(ndt_result.hessian);
+    estimated_cov_2d =
+      pclomp::adjust_diagonal_covariance(cov_by_la, ndt_result.pose, 0.0225, 0.0225);
+  } else if (
+    param_.covariance.covariance_estimation.covariance_estimation_type ==
+    CovarianceEstimationType::MULTI_NDT) {
+    const std::vector<Eigen::Matrix4f> poses_to_search = pclomp::propose_poses_to_search(
+      ndt_result, param_.covariance.covariance_estimation.initial_pose_offset_model_x,
+      param_.covariance.covariance_estimation.initial_pose_offset_model_y);
+    const pclomp::ResultOfMultiNdtCovarianceEstimation result_of_multi_ndt_covariance_estimation =
+      estimate_xy_covariance_by_multi_ndt(ndt_result, ndt_ptr_, poses_to_search);
+    estimated_cov_2d = pclomp::adjust_diagonal_covariance(
+      result_of_multi_ndt_covariance_estimation.covariance, ndt_result.pose, 0.0225, 0.0225);
+    for (const auto & sub_initial_pose_matrix : poses_to_search) {
+      auto sub_output_cloud = std::make_shared<pcl::PointCloud<PointSource>>();
+      ndt_ptr_->align(*sub_output_cloud, sub_initial_pose_matrix);
+      const Eigen::Matrix4f sub_ndt_result = ndt_ptr_->getResult().pose;
+      multi_ndt_result_msg.poses.push_back(matrix4f_to_pose(sub_ndt_result));
+      multi_initial_pose_msg.poses.push_back(matrix4f_to_pose(sub_initial_pose_matrix));
+    }
+    multi_ndt_pose_pub_->publish(multi_ndt_result_msg);
+    multi_initial_pose_pub_->publish(multi_initial_pose_msg);
+  } else if (
+    param_.covariance.covariance_estimation.covariance_estimation_type ==
+    CovarianceEstimationType::MULTI_NDT_SCORE) {
+    const double temperature = 0.1;
```

Comment:
Can this variable be parameterized to config?

Reply:
OK.
I will change the code to handle temperature as a parameter.

Reply:
OK!
[Here](https://github.com/SaltUhey/autoware.universe/commit/8249a12ba22f5058f890df6a71e1c8def76e4df4), I added temperature to the parameters.

Reply:
Thanks!